### PR TITLE
Reconcile request synchronously

### PR DESCRIPTION
**Pull request by sebbonnet**
This changes the reconciliation logic to perform cluster updates synchronously for the same cluster while continuing to allow concurrent updates for different clusters.

Reconciliation for watched resources is done by determining the changes for all resources in one go, so that the delta is calculated at a given point in time (as much as possible). Once all changes are known, we apply them one after the other without stopping on error to try bringing the cluster to the desired state as much as possible.

This work was done on the back of https://github.com/kubernetes-sigs/controller-runtime/issues/616 where we learnt that the reconciler internal queue prevents concurrent updates for the same resource name.
 
Required by #173

### CODE REVIEW

**Comment by sebbonnet** [comment]
As an example here is the list of events generated for a 3-node cluster (in `creation_test.go`).
We can see that multiple nested ADD_SERVICE and ADD_CLUSTER events are dispatched for the same cluster
**Before the changes**
We can see that multiple nested ADD_SERVICE and ADD_CLUSTER events are dispatched for the same cluster. See [cassandras](https://github.com/sky-uk/cassandra-operator/files/3677806/cassandras-with-dispatcher.log) in [operator-logs](https://github.com/sky-uk/cassandra-operator/files/3677807/operator-logs-with-dispatcher.log)

```
time="2019-10-01T16:34:10Z" level=debug msg="Event dispatched id=2 key=test-cassandra-operator.mycluster-g75yx type=ADD_SERVICE"
    time="2019-10-01T16:34:10Z" level=debug msg="Event dispatched id=3 key=test-cassandra-operator.mycluster-g75yx type=ADD_CLUSTER"
    time="2019-10-01T16:34:10Z" level=debug msg="Event received" id=2 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_SERVICE
  time="2019-10-01T16:34:10Z" level=debug msg="Event complete" id=2 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_SERVICE
  time="2019-10-01T16:34:10Z" level=debug msg="Event received" id=3 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CLUSTER
    time="2019-10-01T16:34:10Z" level=debug msg="Event dispatched id=9 key=test-cassandra-operator.mycluster-g75yx type=ADD_CLUSTER"
      time="2019-10-01T16:34:10Z" level=debug msg="Event dispatched id=12 key=test-cassandra-operator.mycluster-g75yx type=ADD_CLUSTER"
        time="2019-10-01T16:34:10Z" level=debug msg="Event dispatched id=14 key=test-cassandra-operator.mycluster-g75yx type=ADD_CUSTOM_CONFIG"
          time="2019-10-01T16:34:10Z" level=debug msg="Event dispatched id=15 key=test-cassandra-operator.mycluster-g75yx type=UPDATE_CLUSTER"
  time="2019-10-01T16:35:08Z" level=debug msg="Event dispatched id=16 key=test-cassandra-operator.mycluster-g75yx type=ADD_CUSTOM_CONFIG"
    time="2019-10-01T16:35:08Z" level=debug msg="Event dispatched id=17 key=test-cassandra-operator.mycluster-g75yx type=UPDATE_CLUSTER"
  time="2019-10-01T16:36:04Z" level=debug msg="Event dispatched id=18 key=test-cassandra-operator.mycluster-g75yx type=ADD_CUSTOM_CONFIG"
    time="2019-10-01T16:36:04Z" level=debug msg="Event dispatched id=19 key=test-cassandra-operator.mycluster-g75yx type=UPDATE_CLUSTER"
time="2019-10-01T16:37:10Z" level=debug msg="Event complete" id=3 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T16:37:10Z" level=debug msg="Event received" id=9 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T16:37:10Z" level=debug msg="Event complete" id=9 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T16:37:10Z" level=debug msg="Event received" id=12 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T16:37:10Z" level=debug msg="Event complete" id=12 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T16:37:10Z" level=debug msg="Event received" id=14 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-01T16:38:40Z" level=debug msg="Event complete" id=14 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-01T16:38:40Z" level=debug msg="Event received" id=15 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=UPDATE_CLUSTER
time="2019-10-01T16:38:40Z" level=debug msg="Event complete" id=15 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=UPDATE_CLUSTER
time="2019-10-01T16:38:40Z" level=debug msg="Event received" id=16 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-01T16:40:10Z" level=debug msg="Event complete" id=16 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-01T16:40:10Z" level=debug msg="Event received" id=17 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=UPDATE_CLUSTER
time="2019-10-01T16:40:10Z" level=debug msg="Event complete" id=17 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=UPDATE_CLUSTER
time="2019-10-01T16:40:10Z" level=debug msg="Event received" id=18 key=test-cassandra-operator.mycluster-g75yx logger=receiver.go type=ADD_CUSTOM_CONFIG
```

**After the changes**
Events are processed one at a time. The number of events is greatly reduced as a result. See [cassandras](https://github.com/sky-uk/cassandra-operator/files/3677808/cassandras-with-parallel-sync-dispatcher.log) in [operator-logs](https://github.com/sky-uk/cassandra-operator/files/3677809/operator-logs-with-parallel-sync-dispatcher.log)

```
 time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=20 key=test-cassandra-operator.mycluster-or8dt type=ADD_SERVICE"
  time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=20 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event complete" id=20 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_SERVICE
  time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=21 key=test-cassandra-operator.mycluster-or8dt type=ADD_CLUSTER"
  time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=21 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T17:20:00Z" level=debug msg="Event complete" id=21 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_CLUSTER
```

Cluster events for different resource names continue to be processed in parallel when the `MaxConcurrentReconciles` is increased
```
time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=16 key=test-cassandra-operator.mycluster-3p94o type=ADD_SERVICE"
time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=16 key=test-cassandra-operator.mycluster-3p94o logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event complete" id=16 key=test-cassandra-operator.mycluster-3p94o logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=17 key=test-cassandra-operator.mycluster-3p94o type=ADD_CLUSTER"
time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=17 key=test-cassandra-operator.mycluster-3p94o logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=18 key=test-cassandra-operator.mycluster-76m6c type=ADD_SERVICE"
time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=18 key=test-cassandra-operator.mycluster-76m6c logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event complete" id=18 key=test-cassandra-operator.mycluster-76m6c logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=19 key=test-cassandra-operator.mycluster-76m6c type=ADD_CLUSTER"
time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=19 key=test-cassandra-operator.mycluster-76m6c logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=20 key=test-cassandra-operator.mycluster-or8dt type=ADD_SERVICE"
time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=20 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event complete" id=20 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_SERVICE
time="2019-10-01T17:19:15Z" level=debug msg="Event dispatched id=21 key=test-cassandra-operator.mycluster-or8dt type=ADD_CLUSTER"
time="2019-10-01T17:19:15Z" level=debug msg="Event received" id=21 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T17:20:00Z" level=debug msg="Event complete" id=21 key=test-cassandra-operator.mycluster-or8dt logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T17:20:15Z" level=debug msg="Event complete" id=19 key=test-cassandra-operator.mycluster-76m6c logger=receiver.go type=ADD_CLUSTER
time="2019-10-01T17:22:30Z" level=debug msg="Event complete" id=17 key=test-cassandra-operator.mycluster-3p94o logger=receiver.go type=ADD_CLUSTER
```

**Comment by alangibson01** [comment]
If it works, and it looks like it does, then it simplifies matters an awful lot. I think there's a good argument for removing the Dispatcher completely now, and just folding the few things it does into the reconciler.

**Comment by sebbonnet** [comment]
I've now removed the dispatcher and change the operations to return an error on failure, so that the reconcile request is retried.
Interestingly `external_modification_test.go` is now failing, as our logic tries to add the custom configmap to the statefulset that was intentionally deleted and fails to do that as the statefulset doesn't exists. This shows that we should probably not return right away when the configmap reconciliation fails, but instead attempt to reconcile the other resources such as statefulset, service, and so on. 
See below extract from the [build logs](https://github.com/sky-uk/cassandra-operator/files/3682797/log.txt)
```
time="2019-10-02T17:45:42Z" level=debug msg="Custom config hash looked up: map[mycluster-un80u-b:6754a6771be9bf2b659e2274de693ec888c7925662957f50fbd9b4e38aa6210d]" cluster=test-cassandra-operator.mycluster-un80u logger=reconciler.go
time="2019-10-02T17:45:42Z" level=debug msg="Custom config is missing on one or more racks. Custom config hashes: map[mycluster-un80u-b:6754a6771be9bf2b659e2274de693ec888c7925662957f50fbd9b4e38aa6210d]. Going to add the configuration" cluster=test-cassandra-operator.mycluster-un80u logger=reconciler.go
time="2019-10-02T17:45:42Z" level=debug msg="Event received" key=test-cassandra-operator.mycluster-un80u logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-02T17:45:42Z" level=info msg="Event will trigger 1 operations: [add custom config for cluster test-cassandra-operator.mycluster-un80u]" key=test-cassandra-operator.mycluster-un80u logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-02T17:45:42Z" level=debug msg="Executing operation add custom config for cluster test-cassandra-operator.mycluster-un80u" key=test-cassandra-operator.mycluster-un80u logger=receiver.go type=ADD_CUSTOM_CONFIG
time="2019-10-02T17:45:42Z" level=info msg="Applying update for rack a in cluster test-cassandra-operator.mycluster-un80u"
time="2019-10-02T17:45:42Z" level=error msg="Error while execution operation add custom config for cluster test-cassandra-operator.mycluster-un80u: unable to add custom configMap to statefulSet for rack a in cluster test-cassandra-operator.mycluster-un80u: unable to retrieve statefulSet for rack a: statefulsets.apps \"mycluster-un80u-a\" not found. Other racks will not be updated. Other racks will not be updated" key=test-cassandra-operator.mycluster-un80u logger=receiver.go type=ADD_CUSTOM_CONFIG
```

**Comment by sebbonnet** [comment]
After changing the reconciler logic to perform all reconciliation in one pass, we're now running into another issue where the Cassandra resource is deleted before the next resource is reconciled. 

Here is the scenario:
- a cassandra cluster is created
- the cluster configmap is changed
- the reconciler detects the configmap change and triggers UPDATE_CUSTOM_CONFIGMAP
- the test finds the change applied and invokes the tear down sequence
  - the cassandra cluster is deleted (via the test). A new reconcile request is created, but queued as a a request for the same cluster is still being processed
  - the statefulset, pvs and service are deleted
- the reconciler carries on processing the same request and now attempts to reconcile the service. At this point the service no longer exists, so an ADD_SERVICE event is triggered
- the reconciler carries on processing the same request (still) and now attempts to reconcile the cluster. At this point the cluster no longer exists, so an ADD_CLUSTER event is triggered which creates the statefulset and pvc again
- the reconciler completes processing the request
- the next reconcile request for the delete cluster is dequeued and processed. The cluster is deleted.
- subsequent tests fail as there are no pv available to bind to.

**Comment by sebbonnet** [comment]
@alangibson01 this is finally ready for review :)

File: cassandra-operator/pkg/operator/reconciler.go [hunk 1]
```diff
@@ -121,7 +116,18 @@ func (r *CassandraReconciler) Reconcile(request reconcile.Request) (reconcile.Re
 	return completeReconciliation()
 }
 
-func (r *CassandraReconciler) reconcileCassandraDefinition(ctx *requestContext, desiredCassandra *v1alpha1.Cassandra) error {
+func (r *CassandraReconciler) appendChanges(determineChange func(*requestContext, *v1alpha1.Cassandra) (*event.Event, error), ctx *requestContext, desiredCassandra *v1alpha1.Cassandra, events []*event.Event, errors *multierror.Error) ([]*event.Event, *multierror.Error) {
```
Comment (alangibson01) [inline cassandra-operator/pkg/operator/reconciler.go]: Might be a bit more readable to define a `type` to use in place of `func(*requestContext, *v1alpha1.Cassandra) (*event.Event, error)`.

**Pull request merged**